File: src/u2ctl/runtime/adb.py
"""ADB discovery, device selection, and recovery operations."""

import shutil
import subprocess
from typing import List, Optional, Tuple

import adbutils

from u2ctl.errors import (
    ADBUnavailableError,
    DeviceNoneError,
    DeviceAmbiguousError,
    DeviceNotFoundError,
    DeviceOfflineError,
    DeviceUnauthorizedError,
)
from u2ctl.models import DeviceInfo


def get_adb_path(custom_adb_path: Optional[str] = None) -> str:
    path = custom_adb_path or shutil.which("adb")
    if not path:
        raise ADBUnavailableError("`adb` command not found on system PATH")
    return path


def list_adb_devices(adb_path: Optional[str] = None) -> List[DeviceInfo]:
    """List all ADB devices with state and transport."""
    path = get_adb_path(adb_path)
    try:
        res = subprocess.run([path, "devices", "-l"], capture_output=True, text=True, check=True)
    except Exception as e:
        raise ADBUnavailableError(f"Failed to execute adb devices: {e}")

    devices = []
    lines = res.stdout.strip().splitlines()
    for line in lines[1:]:  # skip 'List of devices attached' header
        line = line.strip()
        if not line:
            continue
        parts = line.split()
        if len(parts) < 2:
            continue
        serial = parts[0]
        state = parts[1]
        model = ""
        transport = "wifi" if ":" in serial else "usb"

        for p in parts[2:]:
            if p.startswith("model:"):
                model = p.split(":", 1)[1]

        devices.append(DeviceInfo(serial=serial, state=state, model=model, transport=transport))
    return devices


def select_target_device(
    requested_serial: Optional[str] = None,
    adb_path: Optional[str] = None,
) -> Tuple[DeviceInfo, List[DeviceInfo]]:
    """Select target device deterministically according to G8 selection rules."""
    devices = list_adb_devices(adb_path)

    if not devices:
        raise DeviceNoneError("No ADB devices connected")

    if requested_serial:
        matched = [d for d in devices if d.serial == requested_serial]
        if not matched:
            raise DeviceNotFoundError(f"Requested device serial '{requested_serial}' not found in ADB device list")
        target = matched[0]
    else:
        if len(devices) == 1:
            target = devices[0]
        else:
            serials = ", ".join(d.serial for d in devices)
            raise DeviceAmbiguousError(f"Multiple ADB devices connected ({serials}). Must specify --serial")

    # Mark selected
    for d in devices:
        if d.serial == target.serial:
            d.selected = True

    # Validate state of selected device
    if target.state == "unauthorized":
        raise DeviceUnauthorizedError(f"Device '{target.serial}' is unauthorized. Accept RSA key on device screen")
    elif target.state == "offline":
        raise DeviceOfflineError(f"Device '{target.serial}' is offline")
    elif target.state != "device":
        raise DeviceOfflineError(f"Device '{target.serial}' is in non-usable state '{target.state}'")

    return target, devices


def reconnect_device(serial: str, hard: bool = False, adb_path: Optional[str] = None) -> str:
    """Reconnect device via soft adb reconnect or hard server restart."""
    path = get_adb_path(adb_path)
    if hard:
        subprocess.run([path, "kill-server"], capture_output=True, text=True)
        subprocess.run([path, "start-server"], capture_output=True, text=True)
        return "Hard reconnect performed: adb server restarted"
    else:
        res = subprocess.run([path, "reconnect", serial], capture_output=True, text=True)
        return res.stdout.strip() or f"Soft reconnect sent to {serial}"

File: u2bun/src/runtime/adb.ts
import { existsSync } from "node:fs";
import {
  ADBUnavailableError,
  DeviceNoneError,
  DeviceAmbiguousError,
  DeviceNotFoundError,
  DeviceOfflineError,
  DeviceUnauthorizedError,
} from "../errors";
import type { DeviceInfo } from "../models";

export function getAdbPath(customAdbPath?: string): string {
  if (customAdbPath && existsSync(customAdbPath)) {
    return customAdbPath;
  }

  // Check PATH or standard Windows locations
  const whichResult = Bun.which("adb");
  if (whichResult) return whichResult;

  const envPath = process.env.ADB_PATH;
  if (envPath && existsSync(envPath)) return envPath;

  throw new ADBUnavailableError("`adb` command not found on system PATH. Install platform-tools or set ADB_PATH");
}

export async function execAdb(
  args: string[],
  customAdbPath?: string
): Promise<{ stdout: string; stderr: string; exitCode: number }> {
  const adbPath = getAdbPath(customAdbPath);
  const proc = Bun.spawn([adbPath, ...args], {
    stdout: "pipe",
    stderr: "pipe",
    windowsHide: true,
  });

  const stdout = await new Response(proc.stdout).text();
  const stderr = await new Response(proc.stderr).text();
  const exitCode = await proc.exited;

  return { stdout, stderr, exitCode };
}

export async function listAdbDevices(customAdbPath?: string): Promise<DeviceInfo[]> {
  const { stdout, exitCode } = await execAdb(["devices", "-l"], customAdbPath);
  if (exitCode !== 0) {
    throw new ADBUnavailableError("Failed to execute `adb devices -l`");
  }

  const devices: DeviceInfo[] = [];
  const lines = stdout.trim().split("\n");

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;

    const parts = line.split(/\s+/);
    if (parts.length < 2) continue;

    const serial = parts[0];
    const state = parts[1];
    let model = "";
    const transport = serial.includes(":") ? "wifi" : "usb";

    for (let j = 2; j < parts.length; j++) {
      if (parts[j].startsWith("model:")) {
        model = parts[j].split(":", 2)[1];
      }
    }

    devices.push({ serial, state, model, transport });
  }

  return devices;
}

export async function selectTargetDevice(
  requestedSerial?: string,
  customAdbPath?: string
): Promise<{ target: DeviceInfo; devices: DeviceInfo[] }> {
  const devices = await listAdbDevices(customAdbPath);

  if (devices.length === 0) {
    throw new DeviceNoneError("No ADB devices connected");
  }

  let target: DeviceInfo;
  if (requestedSerial) {
    const matched = devices.filter((d) => d.serial === requestedSerial);
    if (matched.length === 0) {
      throw new DeviceNotFoundError(requestedSerial);
    }
    target = matched[0];
  } else {
    if (devices.length === 1) {
      target = devices[0];
    } else {
      const serials = devices.map((d) => d.serial).join(", ");
      throw new DeviceAmbiguousError(serials);
    }
  }

  // Mark selected
  for (const d of devices) {
    if (d.serial === target.serial) {
      d.selected = true;
    }
  }

  // Validate state
  if (target.state === "unauthorized") {
    throw new DeviceUnauthorizedError(target.serial);
  } else if (target.state === "offline") {
    throw new DeviceOfflineError(target.serial);
  } else if (target.state !== "device") {
    throw new DeviceOfflineError(target.serial);
  }

  return { target, devices };
}

export async function reconnectDevice(
  serial: string,
  hard: boolean = false,
  customAdbPath?: string
): Promise<string> {
  if (hard) {
    await execAdb(["kill-server"], customAdbPath);
    await execAdb(["start-server"], customAdbPath);
    return "Hard reconnect performed: adb server restarted";
  } else {
    const { stdout } = await execAdb(["reconnect", serial], customAdbPath);
    return stdout.trim() || `Soft reconnect sent to ${serial}`;
  }
}

export async function forwardPort(
  serial: string,
  localPort: number,
  remotePort: number,
  customAdbPath?: string
): Promise<void> {
  const { exitCode, stderr } = await execAdb(
    ["-s", serial, "forward", `tcp:${localPort}`, `tcp:${remotePort}`],
    customAdbPath
  );
  if (exitCode !== 0) {
    throw new ADBUnavailableError(`Failed to forward port tcp:${localPort} to tcp:${remotePort}: ${stderr}`);
  }
}

export async function inputTextViaAdbKeyboard(
  serial: string,
  text: string,
  customAdbPath?: string
): Promise<boolean> {
  const b64 = Buffer.from(text, "utf8").toString("base64");
  const { stdout } = await execAdb(
    ["-s", serial, "shell", "am", "broadcast", "-a", "ADB_KEYBOARD_INPUT_TEXT", "--es", "text", b64],
    customAdbPath
  );
  // The uiautomator2 AdbKeyboard receiver sets result=-1 on success; without a
  // registered receiver the broadcast completes with result=0 (no dispatch).
  const ok = stdout.includes("result=-1");
  if (ok) {
    await execAdb(
      ["-s", serial, "shell", "am", "broadcast", "-a", "ADB_KEYBOARD_HIDE"],
      customAdbPath
    );
  }
  return ok;
}
